File: packages/sdk-php/src/AI/OpenAIProvider.php
<?php

declare(strict_types=1);

namespace PocketPing\AI;

use PocketPing\Http\CurlHttpClient;
use PocketPing\Http\HttpClientInterface;
use PocketPing\Models\Message;
use PocketPing\Models\Sender;

/**
 * AI provider backed by the OpenAI Chat Completions API.
 */
final class OpenAIProvider implements AIProviderInterface
{
    public const DEFAULT_BASE_URL = 'https://api.openai.com/v1';
    public const DEFAULT_MODEL = 'gpt-4o-mini';

    private readonly string $apiKey;
    private readonly string $model;
    private readonly string $baseUrl;
    private readonly HttpClientInterface $httpClient;

    public function __construct(
        string $apiKey,
        string $model = self::DEFAULT_MODEL,
        string $baseUrl = self::DEFAULT_BASE_URL,
        ?HttpClientInterface $httpClient = null,
    ) {
        $this->apiKey = $apiKey;
        $this->model = $model;
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->httpClient = $httpClient ?? new CurlHttpClient();
    }

    public function name(): string
    {
        return 'openai';
    }

    public function generateResponse(array $messages, ?string $systemPrompt = null): string
    {
        $chatMessages = [];

        if ($systemPrompt !== null && $systemPrompt !== '') {
            $chatMessages[] = ['role' => 'system', 'content' => $systemPrompt];
        }

        foreach ($messages as $msg) {
            $role = $msg->sender === Sender::VISITOR ? 'user' : 'assistant';
            $chatMessages[] = ['role' => $role, 'content' => $msg->content];
        }

        $response = $this->httpClient->post(
            "{$this->baseUrl}/chat/completions",
            [
                'model' => $this->model,
                'messages' => $chatMessages,
                'max_tokens' => 1000,
                'temperature' => 0.7,
            ],
            [
                'Authorization' => "Bearer {$this->apiKey}",
            ],
        );

        if ($response['error'] !== null || $response['httpCode'] < 200 || $response['httpCode'] >= 300) {
            throw new \RuntimeException(
                sprintf('OpenAI API error: %s', $response['error'] ?? "HTTP {$response['httpCode']}")
            );
        }

        $data = json_decode((string) $response['body'], true);
        if (!is_array($data)) {
            return '';
        }

        return (string) ($data['choices'][0]['message']['content'] ?? '');
    }

    public function isAvailable(): bool
    {
        if ($this->apiKey === '') {
            return false;
        }

        $response = $this->httpClient->get(
            "{$this->baseUrl}/models",
            ['Authorization' => "Bearer {$this->apiKey}"],
        );

        return $response['error'] === null
            && $response['httpCode'] >= 200
            && $response['httpCode'] < 300;
    }
}

File: packages/sdk-ruby/lib/pocketping/ai/openai_provider.rb
# frozen_string_literal: true

require_relative "base"

module PocketPing
  module AI
    # OpenAI Chat Completions provider.
    #
    # @example
    #   provider = PocketPing::AI::OpenAIProvider.new(api_key: ENV["OPENAI_API_KEY"])
    #   client = PocketPing::Client.new(ai_provider: provider)
    class OpenAIProvider < Base
      DEFAULT_MODEL = "gpt-4o-mini"
      DEFAULT_BASE_URL = "https://api.openai.com/v1"

      # @param api_key [String] OpenAI API key (required)
      # @param model [String] Model name (default: "gpt-4o-mini")
      # @param base_url [String] API base URL (default: "https://api.openai.com/v1")
      def initialize(api_key:, model: DEFAULT_MODEL, base_url: DEFAULT_BASE_URL)
        super()
        raise ArgumentError, "api_key is required" if api_key.nil? || api_key.empty?

        @api_key = api_key
        @model = model
        @base_url = base_url.chomp("/")
      end

      # @return [String] "openai"
      def name
        "openai"
      end

      # @param messages [Array<PocketPing::Message>]
      # @param system_prompt [String, nil]
      # @return [String]
      def generate_response(messages, system_prompt = nil)
        chat_messages = []
        chat_messages << { role: "system", content: system_prompt } if system_prompt
        messages.each do |message|
          chat_messages << { role: role_for(message), content: message.content }
        end

        body = {
          model: @model,
          messages: chat_messages,
          max_tokens: 1000,
          temperature: 0.7
        }

        data = post_json("#{@base_url}/chat/completions", body, auth_headers)
        data.dig("choices", 0, "message", "content") || ""
      end

      # Check availability by listing models.
      #
      # @return [Boolean]
      def available?
        response = get("#{@base_url}/models", auth_headers)
        response.is_a?(Net::HTTPSuccess)
      rescue StandardError
        false
      end

      private

      def auth_headers
        { "Authorization" => "Bearer #{@api_key}" }
      end
    end
  end
end
